sycl/source/detail/os_util.cpp

```diff
@@ -77,6 +87,99 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
   return reinterpret_cast<OSModuleHandle>(Res.Handle);
 }
 
+bool procMapsAddressInRange(std::istream &Stream, uintptr_t Addr) {
+  uintptr_t Start, End;
+  Stream >> Start;
+  assert(!Stream.fail() && Stream.peek() == '-' &&
+         "Couldn't read /proc/self/maps correctly");
+  Stream.ignore(1);
+
+  Stream >> End;
+  assert(!Stream.fail() && Stream.peek() == ' ' &&
+         "Couldn't read /proc/self/maps correctly");
+  Stream.ignore(1);
+
+  return Addr >= Start && Addr < End;
+}
+
+/// Returns an absolute path to a directory where the object was found.
+std::string OSUtil::getCurrentDSODir() {
+  // Examine /proc/self/maps and find where ^~~this function comes from - this
+  // is supposed to be an absolute path to libsycl.so.
+  //
+  // File structure is the following:
+  //   address           perms offset  dev   inode       pathname
+  //   00400000-00452000 r-xp 00000000 08:02 173521      /usr/bin/foo
+  //   007c2000-007c8000 r--p 001c2000 fc:05 52567930    /usr/bin/bar
+  //
+  // We need to:
+  //
+  //  1) Iterate over lines and find the line which have an address of the
+  //     current function in an `address' range.
+  //
+  //  2) Check that perms have read and executable flags (since we do execute
+  //     this function).
+  //
+  //  3) Skip offset, dev, inode
+  //
+  //  4) Extract an absolute path to a filename and get a dirname from it.
+  //
+  uintptr_t CurrentFunc = (uintptr_t) &getCurrentDSODir;
+  std::ifstream Stream("/proc/self/maps");
+  Stream >> std::hex;
+  while (!Stream.eof()) {
+    if (!procMapsAddressInRange(Stream, CurrentFunc)) {
+      // Skip the rest until an EOL and check the next line
+      Stream.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+      continue;
+    }
+
+    char Perm[4];
+    Stream.readsome(Perm, sizeof(Perm));
+    assert(Perm[0] == 'r' && Perm[2] == 'x' &&
+           "Invalid flags in /proc/self/maps");
+    assert(Stream.peek() == ' ');
+    Stream.ignore(1);
+
+    // Read and ignore the following:
+    uintptr_t Offset, DevMajor, DevMinor, Inode;
+    Stream >> Offset;
+    assert(Stream.peek() == ' ');
+    Stream.ignore(1);
+
+    Stream >> DevMajor;
+    assert(Stream.peek() == ':' &&
+           "Couldn't read dev field in /proc/self/maps");
+    Stream.ignore(1);
+    Stream >> DevMinor;
+    assert(Stream.peek() == ' ');
+    Stream.ignore(1);
+
+    Stream >> Inode;
+    assert(Stream.peek() == ' ');
+    Stream.ignore(1);
+
+    // Now read the path: it is padded with whitespaces, so we skip them
+    // first.
+    while (Stream.peek() == ' ') {
+      Stream.ignore(1);
+    }
+    char Path[PATH_MAX];
+    Stream.getline(Path, PATH_MAX - 1);
+    Path[PATH_MAX - 1] = '\0';
+    return OSUtil::getDirName(Path);
+  }
+  assert(false && "Unable to find the current function in /proc/self/maps");
+  return "";
+}
+
+std::string OSUtil::getDirName(const char* Path) {
+  std::string Tmp(Path);
```

Comment:
Probably I'm missing something, but this function does two copies of a string.
Wouldn't it be better to do something like:
```

const char *LastSlash = nullptr;
while(*Path != '\0')
  if(*Path == '/')
   LastSlash = Path;
return std::string(Path, size_t(LastSlash - Path));
```

?